File: application/static/js/helpers/price.js
/*global define */
/*global _ */

define(['helpers/currency', 'helpers/date', 'lib/underscore'], function (currencyHelper, dateHelper) {
	var specialRepeat = [
		{'value': 'no', 'converter': function (date) {return date; }},
		{'value': 'week', 'converter': function (date) {return date.getDay(); }},
		{'value': 'month', 'converter': function (date) {return date.getDate(); }},
		{'value': 'year', 'converter': dateHelper.getDayOfYear}
	];
	var specialMap = {};
	(function () {
		for (var i = specialRepeat.length - 1; i >= 0; i -= 1) {
			specialMap[specialRepeat[i].value] = specialRepeat[i];
		}
	})();

	var inCircularRange = function (start, end, val) {
		return (start < end && start <= val && val < end) || // eg. start: Tue end: Fri
				(start > end && (val < end || val >= start)); // eg: start: Sat end: Tue
	};

	var checkIfDateBetween = function (converter, startDate, endDate, date) {
		var start = converter.call(null, startDate);
		var end = converter.call(null, endDate);
		var val = converter.call(null, date);
		return inCircularRange(start, end, val);
	};

	var specialAppliesTo = function (special, date) {
		special.startDate = special.startDate || dateHelper.toDate(special.start);
		special.endDate = special.endDate || dateHelper.toDate(special.end);
		var sr = specialMap[special.repeat];
		return checkIfDateBetween(sr.converter, special.startDate, special.endDate, date);
	};

	var getPriceValuesForDate = function (prices, date) {
		var n = prices.special.length;
		for (var i = 0; i < n; i += 1) {
			var special = prices.special[i];
			if (specialAppliesTo(special, date)) {
				console.log('\tapply ' + special.repeat + 'ly (' +
					special.start + ' - ' + special.end + '): ' +
					special.values);
				return special.values;
			}
		}
		return prices.values;
	};

	var getAvgPricesForDays = function (pricesPerDays) {
		var avg = [];
		var guests = pricesPerDays[0].length;
		var days = pricesPerDays.length;
		console.log('price matrix: ' + pricesPerDays);
		for (var guestsIdx = guests - 1; guestsIdx >= 0; guestsIdx -= 1) {
			var avgForGuest = _.reduce(pricesPerDays, function (sum, pricesPerDay) {
				return sum + pricesPerDay[guestsIdx];
			}, 0) / days;
			avg.unshift(avgForGuest);
		}

		console.log('avgs: ' + avg);
		return avg;
	};

	var getAvgForRange = function (prices, start, end) {
		if (prices.special && prices.special.length > 0 && dateHelper.isValidRange(start, end)) {
			var pricesPerDays = []; // will contain prices for each day
			dateHelper.iterateBetweenDates(start, end, 'date', function (date) {
				// for every date, retrieve price values
				console.log('for date: ' + dateHelper.toStr(date));
				var vals = getPriceValuesForDate(prices, date);
				console.log('\twith vals: ' + vals);
				pricesPerDays.push(vals);
			});
			return getAvgPricesForDays(pricesPerDays);
		}
		return prices.values;
	};

	var calcPerNight = function (prices, quantity, guests, places, start, end) {
		var vals = getAvgForRange(prices, start, end);
		var p = places;
		var price = vals[0] * quantity; // every room should have at least 1 guest
		guests = Math.max(0, guests - quantity); // calculate without these guests
		var f = p > 1 ? Math.floor(guests / (p - 1)):quantity; // nr of full rooms
		var rg = p  > 1 ? guests % (p - 1):0; // nr of guests that are not in full rooms
		price = price - f * vals[0] + f * vals[p - 1]; // add prices of full rooms
		price = price - vals[0] + vals[rg]; // add price of partially filled room
		return currencyHelper.convertDefaultTo(price);
	};

	return {
		'specialRepeat': specialRepeat,
		'calcPerNight': calcPerNight
	};
});

File: application/helpers/price.py
from application.helpers import date as date_helper
import math
import logging

special_map = {
    'no': {'converter': lambda date: date},
    'week': {'converter': lambda date: date},
    'month': {'converter': lambda date: date},
    'year': {'converter': lambda date: date},
}


def in_circular_range(start, end, val):
    # eg. start: Tue end: Fri
    # or
    # eg: start: Sat end: Tue
    return ((start < end and start <= val and val < end) or
            (start > end and (val < end or val >= start)))


def check_if_date_between(converter, start_date, end_date, date):
    start = converter(start_date)
    end = converter(end_date)
    val = converter(date)
    return in_circular_range(start, end, val)


def special_applies_to(special, date):
    if 'start_date' not in special:
        special['start_date'] = date_helper.to_date(special['start'])
    if 'end_date' not in special:
        special['end_date'] = date_helper.to_date(special['end'])
    sr = special_map[special['repeat']]
    return check_if_date_between(
        sr['converter'],
        special['start_date'],
        special['end_date'], date)


def get_price_values_for_date(prices, date):
    n = len(prices['special'])
    for i in range(0, n):
        special = prices['special'][i]
        if special_applies_to(special, date):
            logging.info(
                '\tapply ' + special['repeat'] + 'ly (' +
                special['start'] + ' - ' + special['end'] + '): ' +
                str(special['values']))
            return special['values']

    return prices['values']


def get_avg_prices_for_days(prices_per_days):
    avg = []
    guests = len(prices_per_days[0])
    days = len(prices_per_days)
    logging.info('price matrix: ' + str(prices_per_days))
    for guests_idx in range(0, guests):
        avg_for_guest = reduce(
            lambda sm, price_per_day: sm + price_per_day[guests_idx],
            prices_per_days,
            0) / days
        avg += [avg_for_guest]

    logging.info('avgs: ' + str(avg))
    return avg


def get_avg_for_range(prices, start, end):
    if prices['special']\
            and len(prices['special']) > 0\
            and start < end:
        prices_per_days = []  # will contain prices for each day
        for date in date_helper.date_range(start, end):
            # for every date, retrieve price values
            vals = get_price_values_for_date(prices, date)
            logging.info('\twith vals: ' + str(vals))
            logging.info('for date: ' + str(date_helper.to_str(date)))
            prices_per_days += [vals]

        return get_avg_prices_for_days(prices_per_days)

    return prices['values']


def get_price_for(prices, quantity, guests, start, end):
    vals = get_avg_for_range(prices, start, end)
    places = len(prices['values'])
    # every room should have at least 1 guest
    price_per_day = vals[0] * quantity
    # calculate without these guests
    guests = max(0, guests - quantity)
    # nr of full rooms
    f = int(guests / (places - 1)) if places > 1 else quantity
    # nr of guests that are not in full rooms
    rg = guests % (places - 1) if places > 1 else 0
    # add prices of full rooms
    price_per_day = price_per_day - f * vals[0] + f * vals[places - 1]
    # add price_per_day of partially filled room
    price_per_day = price_per_day - vals[0] + vals[rg]

    price_per_day = math.ceil(price_per_day)

    days = (end - start).days
    return price_per_day * days
